File: src/sarma_cli/engine/agent_factory.py
"""LangGraph ReAct Agent factory."""

from __future__ import annotations

import json
import logging
from dataclasses import asdict
from typing import Any

from sarma_cli.engine.errors import AgentBuildError, ProviderNotConfiguredError
from sarma_cli.engine.mcp_pool import McpClientPool
from sarma_cli.engine.models import AgentRunConfig, ResolvedSkill
from sarma_cli.engine.model_factory import ModelFactory
from sarma_cli.runtime.middleware import build_agent_middleware_for_model
from sarma_cli.runtime.services import AgentRuntimeServices
from sarma_cli.resources.network_tools import (
    build_http_exchange_tool,
    build_packet_exchange_tool,
)
from sarma_cli.resources.rag import build_rag_search_tool
from sarma_cli.resources.web_tools import build_web_search_tool

logger = logging.getLogger(__name__)


def _provider_key(provider: Any) -> dict[str, Any]:
    if hasattr(provider, "to_dict"):
        return provider.to_dict()
    return {
        "name": getattr(provider, "name", ""),
        "model_name": getattr(provider, "model_name", ""),
        "api_mode": getattr(provider, "api_mode", ""),
        "api_key": getattr(provider, "api_key", ""),
        "base_url": getattr(provider, "base_url", ""),
        "temperature": getattr(provider, "temperature", None),
        "top_p": getattr(provider, "top_p", None),
    }


def _skill_key(skill: ResolvedSkill | None) -> dict[str, Any] | None:
    if skill is None:
        return None
    return {
        "id": skill.id,
        "name": skill.name,
        "system_prompt_suffix": skill.system_prompt_suffix,
        "tool_allowlist": None if skill.tool_allowlist is None else sorted(skill.tool_allowlist),
        "tool_denylist": None if skill.tool_denylist is None else sorted(skill.tool_denylist),
        "preferred_model_name": skill.preferred_model_name,
        "temperature_override": skill.temperature_override,
    }


def _rag_key(rag: Any) -> dict[str, Any]:
    try:
        return asdict(rag)
    except TypeError:
        return {}


class AgentFactory:
    """Builds a LangGraph agent from runtime configuration."""

    def __init__(
        self,
        pool: McpClientPool,
        workspace_path: str = "",
        model_factory: ModelFactory | None = None,
        runtime_services: AgentRuntimeServices | None = None,
    ) -> None:
        self._pool = pool
        self._workspace_path = workspace_path
        self._model_factory = model_factory or ModelFactory()
        self._runtime_services = runtime_services
        self._agent_cache: dict[str, tuple[Any, list[Any]]] = {}
        self._agent_cache_limit = 8

    async def build(
        self, config: AgentRunConfig
    ) -> tuple[Any, list[Any]]:
        """Build and return (compiled_graph, tools).

        Args:
            config: Full run configuration.

        Returns:
            Tuple of (compiled LangGraph agent, list of LangChain tools).

        Raises:
            ProviderNotConfiguredError: If the provider is invalid.
            AgentBuildError: If agent construction fails.
        """
        provider = config.provider

        # Validate provider has minimum fields
        if not provider.model_name:
            raise ProviderNotConfiguredError(
                "Model name is required for the selected provider."
            )

        # 1. Build MCP server configs from enabled servers
        server_configs: dict[str, dict[str, Any]] = {}
        for server in config.enabled_servers:
            server_configs[server.name] = server.to_langchain_config()

        # 2. Connect / reuse MCP client pool and get tools
        all_tools = await self._pool.connect(server_configs)

        # 3. Apply skill tool filter and append built-in local tools.
        tools = self._apply_skill_filter(all_tools, config.skill)
        tools.extend(self._build_builtin_tools(config))

        if not tools:
            logger.info(
                "No MCP tools available (no servers connected or all filtered "
                "out). Agent will answer from its own knowledge."
            )

        cache_key = self._agent_cache_key(config, server_configs, tools)
        cached = self._agent_cache.get(cache_key)
        if cached is not None:
            logger.info(
                "Agent cache hit: mode=%s model=%s tools=%d skill=%s",
                config.mode,
                provider.model_name,
                len(tools),
                config.skill.name if config.skill else "none",
            )
            return cached

        # 4. Initialize LLM
        try:
            model = self._init_model(provider, config.skill)
        except Exception as exc:
            raise AgentBuildError(
                f"Failed to initialize model: {exc}"
            ) from exc

        # 5. Build agent. Audit modes use fixed pipelines; Ruflo uses a
        #    primary ReAct agent with a controlled subagent delegation tool.
        try:
            agent = self._create_agent(config, model, tools)
        except Exception as exc:
            raise AgentBuildError(
                f"Failed to create agent: {exc}"
            ) from exc

        logger.info(
            "Agent built: model=%s, tools=%d, skill=%s",
            provider.model_name,
            len(tools),
            config.skill.name if config.skill else "none",
        )

        self._agent_cache[cache_key] = (agent, tools)
        if len(self._agent_cache) > self._agent_cache_limit:
            self._agent_cache.pop(next(iter(self._agent_cache)))
        return agent, tools

    def _create_agent(
        self,
        config: AgentRunConfig,
        model: Any,
        tools: list[Any],
    ) -> Any:
        if config.mode in ("audit", "audit-slim"):
            subagent_models = self._load_subagent_models(config.subagent_providers)
            subagent_models.pop("orchestrator", None)
            compile_kwargs = (
                self._runtime_services.compile_kwargs()
                if self._runtime_services is not None
                else {}
            )

            if config.mode == "audit-slim":
                from sarma_cli.engine.audit_slim_graph import build_audit_slim_graph
                from sarma_cli.engine.audit_slim_subagents import AUDIT_SLIM_SUBAGENTS

                return build_audit_slim_graph(
                    model=model,
                    tools=tools,
                    system_prompt=config.system_prompt or "",
                    subagent_specs=AUDIT_SLIM_SUBAGENTS,
                    subagent_models=subagent_models or None,
                    subagent_mcp_allow=config.subagent_mcp_allow,
                    subagent_skills=config.subagent_skills,
                    compile_kwargs=compile_kwargs,
                )

            from sarma_cli.engine.audit_graph import build_audit_graph
            from sarma_cli.engine.audit_subagents import AUDIT_SUBAGENTS

            return build_audit_graph(
                model=model,
                tools=tools,
                system_prompt=config.system_prompt or "",
                subagent_specs=AUDIT_SUBAGENTS,
                subagent_models=subagent_models or None,
                subagent_mcp_allow=config.subagent_mcp_allow,
                subagent_skills=config.subagent_skills,
                compile_kwargs=compile_kwargs,
            )

        if config.mode == "ruflo":
            from langchain.agents import create_agent
            from sarma_cli.engine.ruflo import build_delegate_tool, build_ruflo_prompt

            system_prompt = build_ruflo_prompt(config.system_prompt or "")
            ruflo_tools = [*tools, build_delegate_tool(model, tools)]
            agent_kwargs = (
                self._runtime_services.create_agent_kwargs()
                if self._runtime_services is not None
                else {}
            )
            return create_agent(
                model,
                ruflo_tools,
                system_prompt=system_prompt,
                middleware=build_agent_middleware_for_model(model),
                **agent_kwargs,
            )

        from langchain.agents import create_agent

        system_prompt = config.system_prompt or ""
        agent_kwargs = (
            self._runtime_services.create_agent_kwargs()
            if self._runtime_services is not None
            else {}
        )
        return create_agent(
            model,
            tools,
            system_prompt=system_prompt,
            middleware=build_agent_middleware_for_model(model),
            **agent_kwargs,
        )

    def _agent_cache_key(
        self,
        config: AgentRunConfig,
        server_configs: dict[str, dict[str, Any]],
        tools: list[Any],
    ) -> str:
        data = {
            "mode": config.mode,
            "provider": _provider_key(config.provider),
            "skill": _skill_key(config.skill),
            "servers": server_configs,
            "tools": [getattr(tool, "name", repr(tool)) for tool in tools],
            "system_prompt": config.system_prompt or "",
            "subagent_providers": {
                name: _provider_key(provider)
                for name, provider in sorted(config.subagent_providers.items())
            },
            "subagent_mcp_allow": {
                name: None if allow is None else sorted(allow)
                for name, allow in sorted(config.subagent_mcp_allow.items())
            },
            "subagent_skills": {
                name: _skill_key(skill)
                for name, skill in sorted(config.subagent_skills.items())
            },
            "rag": _rag_key(config.rag),
        }
        return json.dumps(data, sort_keys=True, default=str)

    def _init_model(self, provider: Any, skill: ResolvedSkill | None) -> Any:
        """Initialize a LangChain language model based on api_mode."""
        return self._model_factory.init_model(provider, skill)

    def _apply_skill_filter(
        self,
        tools: list[Any],
        skill: ResolvedSkill | None,
    ) -> list[Any]:
        """Filter tools based on skill allow/deny lists."""
        if skill is None:
            return list(tools)
        return list(
            self._pool.filter_tools(
                tools,
                allowlist=skill.tool_allowlist,
                denylist=skill.tool_denylist,
            )
        )

    def _load_subagent_models(self, providers: dict[str, Any]) -> dict[str, Any]:
        models: dict[str, Any] = {}
        for name, provider in providers.items():
            models[name] = self._init_model(provider, None)
        return models

    def _build_builtin_tools(self, config: AgentRunConfig) -> list[Any]:
        tools = [
            build_web_search_tool(),
            build_http_exchange_tool(),
            build_packet_exchange_tool(),
        ]
        if any(kb.enabled and kb.name for kb in config.rag.knowledge_bases):
            tools.append(build_rag_search_tool(config.rag))
        return tools

File: src/engine/agentFactory.ts
/** LangGraph agent factory. */

import { createAgent } from "langchain";
import type { BaseChatModel } from "@langchain/core/language_models/chat_models";
import type { StructuredToolInterface } from "@langchain/core/tools";

import { AgentBuildError, ProviderNotConfiguredError } from "@/engine/errors";
import { McpClientPool } from "@/engine/mcpPool";
import { type AgentRunConfig, ResolvedSkill } from "@/engine/models";
import { ModelFactory } from "@/engine/modelFactory";
import type { ModelProviderDTO } from "@/engine/dto";
import { buildAgentMiddlewareForModel } from "@/runtime/middleware";
import { AgentRuntimeServices } from "@/runtime/services";
import { buildHttpExchangeTool, buildPacketExchangeTool } from "@/resources/networkTools";
import { buildWebSearchTool } from "@/resources/webTools";
import { buildRagSearchTool } from "@/resources/rag";
import { buildAuditGraph } from "@/workflows/auditGraph";
import { AUDIT_SUBAGENTS } from "@/workflows/auditSubagents";
import { buildAuditSlimGraph } from "@/workflows/auditSlimGraph";
import { AUDIT_SLIM_SUBAGENTS } from "@/workflows/auditSlimSubagents";
import { buildDelegateTool } from "@/workflows/ruflo";
import { createTokenEstimator } from "@/context/tokenizer";

type CompiledAgent = { stream: (...args: unknown[]) => AsyncIterable<unknown> } & Record<string, unknown>;
type BuildResult = [CompiledAgent, StructuredToolInterface[]];

function providerKey(provider: ModelProviderDTO): Record<string, unknown> {
  return {
    name: provider.name,
    model_name: provider.modelName,
    api_mode: provider.apiMode,
    api_key: provider.apiKey,
    base_url: provider.baseUrl,
    temperature: provider.temperature,
    top_p: provider.topP,
  };
}

function skillKey(skill: ResolvedSkill | null): Record<string, unknown> | null {
  if (skill === null) return null;
  return {
    id: skill.id,
    name: skill.name,
    system_prompt_suffix: skill.systemPromptSuffix,
    tool_allowlist: skill.toolAllowlist === null ? null : [...skill.toolAllowlist].sort(),
    tool_denylist: skill.toolDenylist === null ? null : [...skill.toolDenylist].sort(),
    preferred_model_name: skill.preferredModelName,
    temperature_override: skill.temperatureOverride,
  };
}

function ragKey(rag: AgentRunConfig["rag"]): Record<string, unknown> {
  return {
    embedding_backend: rag.embeddingBackend,
    embedding_model: rag.embeddingModel,
    embedding_api_base: rag.embeddingApiBase,
    embedding_api_key: rag.embeddingApiKey,
    embedding_local_path: rag.embeddingLocalPath,
    chunk_size: rag.chunkSize,
    chunk_overlap: rag.chunkOverlap,
    knowledge_bases: rag.knowledgeBases.map((kb) => ({
      name: kb.name,
      backend: kb.backend,
      docs_path: kb.docsPath,
      chroma_path: kb.chromaPath,
      chroma_url: kb.chromaUrl,
      collection_name: kb.collectionName,
      tenant: kb.tenant,
      database: kb.database,
      headers: kb.headers,
      enabled: kb.enabled,
    })),
  };
}

/** Builds a LangGraph agent from runtime configuration. */
export class AgentFactory {
  private readonly modelFactory: ModelFactory;
  private readonly runtimeServices: AgentRuntimeServices | null;
  private readonly agentCache = new Map<string, BuildResult>();
  private readonly agentCacheLimit = 8;

  constructor(
    private readonly pool: McpClientPool,
    options: {
      workspacePath?: string;
      modelFactory?: ModelFactory;
      runtimeServices?: AgentRuntimeServices | null;
    } = {},
  ) {
    this.modelFactory = options.modelFactory ?? new ModelFactory();
    this.runtimeServices = options.runtimeServices ?? null;
  }

  /** Build and return [compiledGraph, tools]. */
  async build(config: AgentRunConfig): Promise<BuildResult> {
    const provider = config.provider;
    if (!provider.modelName) {
      throw new ProviderNotConfiguredError("Model name is required for the selected provider.");
    }

    // 1. Build MCP server configs from enabled servers.
    const serverConfigs: Record<string, Record<string, unknown>> = {};
    for (const server of config.enabledServers) {
      serverConfigs[server.name] = server.toLangchainConfig();
    }

    // 2. Connect / reuse MCP client pool and get tools.
    const allTools = await this.pool.connect(serverConfigs);

    // 3. Match Python Sarma: skill allow/deny filters MCP tools, then
    // built-in local tools are appended and remain available.
    const tools = [
      ...this.applySkillFilter(allTools, config.skill),
      ...this.buildBuiltinTools(config),
    ];

    const cacheKey = this.agentCacheKey(config, serverConfigs, tools);
    const cached = this.agentCache.get(cacheKey);
    if (cached !== undefined) return cached;

    // 4. Initialize LLM.
    let model: BaseChatModel;
    try {
      model = this.modelFactory.initModel(provider, config.skill);
    } catch (exc) {
      throw new AgentBuildError(`Failed to initialize model: ${exc instanceof Error ? exc.message : exc}`);
    }

    // 5. Build agent (audit pipelines vs ruflo primary + delegation).
    let agent: CompiledAgent;
    try {
      agent = this.createAgentForMode(config, model, tools);
    } catch (exc) {
      throw new AgentBuildError(`Failed to create agent: ${exc instanceof Error ? exc.message : exc}`);
    }

    const result: BuildResult = [agent, tools];
    this.agentCache.set(cacheKey, result);
    if (this.agentCache.size > this.agentCacheLimit) {
      const firstKey = this.agentCache.keys().next().value;
      if (firstKey !== undefined) this.agentCache.delete(firstKey);
    }
    return result;
  }

  private createAgentForMode(
    config: AgentRunConfig,
    model: BaseChatModel,
    tools: StructuredToolInterface[],
  ): CompiledAgent {
    if (config.mode === "audit" || config.mode === "audit-slim") {
      const subagentModels = this.loadSubagentModels(config.subagentProviders);
      delete subagentModels.orchestrator;
      const compileKwargs = this.runtimeServices?.compileKwargs() ?? {};
      const maxPriorStageTokens = Math.max(
        12_000,
        Math.min(120_000, Math.trunc((config.provider.maxContextTokens || 128_000) * 0.35)),
      );
      const estimateText = createTokenEstimator(config.provider);

      if (config.mode === "audit-slim") {
        return buildAuditSlimGraph(model, tools, {
          systemPrompt: config.systemPrompt || "",
          subagentSpecs: AUDIT_SLIM_SUBAGENTS,
          subagentModels: Object.keys(subagentModels).length ? subagentModels : null,
          subagentMcpAllow: config.subagentMcpAllow,
          subagentSkills: config.subagentSkills,
          maxPriorStageTokens,
          estimateText,
          compileKwargs,
        }) as unknown as CompiledAgent;
      }

      return buildAuditGraph(model, tools, {
        systemPrompt: config.systemPrompt || "",
        subagentSpecs: AUDIT_SUBAGENTS,
        subagentModels: Object.keys(subagentModels).length ? subagentModels : null,
        subagentMcpAllow: config.subagentMcpAllow,
        subagentSkills: config.subagentSkills,
        maxPriorStageTokens,
        estimateText,
        compileKwargs,
      }) as unknown as CompiledAgent;
    }

    if (config.mode === "ruflo") {
      const rufloTools = [...tools, buildDelegateTool(model, tools)];
      const agentKwargs = this.runtimeServices?.createAgentKwargs() ?? {};
      return createAgent({
        model,
        tools: rufloTools,
        systemPrompt: config.systemPrompt || "",
        middleware: buildAgentMiddlewareForModel(model),
        ...agentKwargs,
      }) as unknown as CompiledAgent;
    }

    const agentKwargs = this.runtimeServices?.createAgentKwargs() ?? {};
    return createAgent({
      model,
      tools,
      systemPrompt: config.systemPrompt || "",
      middleware: buildAgentMiddlewareForModel(model),
      ...agentKwargs,
    }) as unknown as CompiledAgent;
  }

  private agentCacheKey(
    config: AgentRunConfig,
    serverConfigs: Record<string, Record<string, unknown>>,
    tools: StructuredToolInterface[],
  ): string {
    const sortedEntries = <T>(obj: Record<string, T>): [string, T][] =>
      Object.entries(obj).sort(([a], [b]) => a.localeCompare(b));

    const data = {
      mode: config.mode,
      provider: providerKey(config.provider),
      skill: skillKey(config.skill),
      servers: serverConfigs,
      tools: tools.map((t) => t.name ?? String(t)),
      system_prompt: config.systemPrompt || "",
      subagent_providers: Object.fromEntries(
        sortedEntries(config.subagentProviders).map(([name, p]) => [name, providerKey(p)]),
      ),
      subagent_mcp_allow: Object.fromEntries(
        sortedEntries(config.subagentMcpAllow).map(([name, allow]) => [
          name,
          allow === null ? null : [...allow].sort(),
        ]),
      ),
      subagent_skills: Object.fromEntries(
        sortedEntries(config.subagentSkills).map(([name, skill]) => [name, skillKey(skill)]),
      ),
      rag: ragKey(config.rag),
    };
    return stableStringify(data);
  }

  private applySkillFilter(
    tools: StructuredToolInterface[],
    skill: ResolvedSkill | null,
  ): StructuredToolInterface[] {
    if (skill === null) return [...tools];
    return [...this.pool.filterTools(tools, skill.toolAllowlist, skill.toolDenylist)];
  }

  private loadSubagentModels(
    providers: Record<string, ModelProviderDTO>,
  ): Record<string, BaseChatModel> {
    const models: Record<string, BaseChatModel> = {};
    for (const [name, provider] of Object.entries(providers)) {
      models[name] = this.modelFactory.initModel(provider, null);
    }
    return models;
  }

  private buildBuiltinTools(config: AgentRunConfig): StructuredToolInterface[] {
    const tools: StructuredToolInterface[] = [
      buildWebSearchTool(),
      buildHttpExchangeTool(),
      buildPacketExchangeTool(),
    ];
    if (config.rag.knowledgeBases.some((kb) => kb.enabled && kb.name)) {
      tools.push(buildRagSearchTool(config.rag));
    }
    return tools;
  }
}

/** Deterministic JSON stringify with sorted object keys (matches Python sort_keys). */
function stableStringify(value: unknown): string {
  return JSON.stringify(sortKeys(value));
}

function sortKeys(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortKeys);
  if (value && typeof value === "object") {
    const out: Record<string, unknown> = {};
    for (const key of Object.keys(value as Record<string, unknown>).sort()) {
      out[key] = sortKeys((value as Record<string, unknown>)[key]);
    }
    return out;
  }
  return value;
}
